packages/datadog-plugin-rhea/src/index.js

```diff
@@ -0,0 +1,146 @@
+        'amqp.link.target.address': name,
+        'amqp.link.role': 'sender',
+        'out.host': this.connection.options.host,
+        'out.port': this.connection.options.port
+      } }, (span, done) => {
+        msg.delivery_annotations = msg.delivery_annotations || {}
+        tracer.inject(span, 'text_map', msg.delivery_annotations)
+        const delivery = send.apply(this, arguments)
+        if (this.options.snd_settle_mode !== 1) {
+          delivery._dd = { done, span }
+        } else {
+          done()
+        }
+        return delivery
+      })
+    }
+  }
+}
+
+function createWrapConnectionDispatch (tracer, config) {
+  return function wrapDispatch (dispatch) {
+    return function dispatchWithTrace (eventName, obj) {
+      if (eventName === 'disconnected') {
+        for (const key in this.local_channel_map) {
+          const session = this.local_channel_map[key]
+          const addTags = entry => {
+            if (entry && entry._dd) {
+              const { span, done } = entry._dd
+              const error = obj.error || this.saved_error
+              if (error) {
+                span.addTags({
+                  'error.type': error.name,
+                  'error.msg': error.message,
+                  'error.stack': error.stack
+                })
+              }
+              done()
+            }
+          }
+          session.incoming.deliveries.entries.forEach(addTags)
+          session.outgoing.deliveries.entries.forEach(addTags)
+        }
+      }
+      return dispatch.apply(this, arguments)
+    }
+  }
+}
+
+function createWrapSenderDispatch (tracer, config) {
+  return function wrapDispatch (dispatch) {
+    return function dispatchWithTrace (eventName, obj) {
+      if (eventName === 'settled') {
+        const state = obj.delivery.remote_state.constructor.composite_type
+        obj.delivery._dd.span.setTag('amqp.delivery.state', state)
+        obj.delivery._dd.done()
+      }
+      return dispatch.apply(this, arguments)
+    }
+  }
+}
+
+function createWrapReceiverDispatch (tracer, config) {
+  return function wrapDispatch (dispatch) {
+    return function dispatchWithTrace (eventName, msgObj) {
+      if (eventName === 'message') {
+        const name = msgObj.receiver.options.source && msgObj.receiver.options.source.address
+          ? msgObj.receiver.options.source.address : 'amq.topic'
+
+        const childOf = tracer.extract('text_map', msgObj.message.delivery_annotations)
+        return tracer.trace('rhea.receiver.onmessage', {
+          tags: {
+            'resource.name': name,
+            'service.name': config.service || `${tracer._service}-amqp`,
+            'span.kind': 'consumer',
+            'amqp.link.source.address': name,
+            'amqp.link.role': 'receiver'
+          },
+          childOf
+        }, (span, done) => {
+          msgObj.delivery._dd = { done, span }
+          wrapDeliveryMethod('accept', 'accepted', msgObj.delivery)
```

Comment:
They do! However, the object that's passed to that common method doesn't have a human-readable version of the intended delivery state. Since delivery state is what we're trying to get at here, it makes sense to wrap the methods individually so we have that information.

Reply:
The problem is that there are places in the code that call `update` directly and it's exposed to the user. It's also possible to extend AMQP with custom states. This means that the only safe approach would be to patch the underlying `update` method.

Reply:
Alright I'll do a mapping from numeric codes to strings. That doesn't handle custom states, but `rhea` doesn't seem to support them to begin with.